### ISSUE TITLE
Transfer: Display selected items that are not provided as options

### ISSUE THREAD

**Issue by martinkrulltott** [dhis2/ui#143]
Based on [TECH-380](https://jira.dhis2.org/browse/TECH-380).

When working with large amounts of data there's a need to paginate and sequentially load more data dynamically. This means that a full dataset cannot be expected at all times and will not always be available as `options.` If an item has been previously selected (`id` passed as `selected`) but isn't currently pass as part of `options` it still needs to be displayed on the right side as a selected item.

Currently, the `Transfer` component filters out `selected` items that are not part of the `options`, which leads to these `selected` items being hidden.

Example from Data Visualizer: When opening a saved analytics object it will contain selected items, but the available dataset for a dimension selector (e.g. Data) that's using the `Transfer` component will not necessarily have loaded that part of the full dataset (`options`). Meaning the current `options` can now be exclusive of the selected item.

**Comment by ismay** [comment]
The Select does something similar, it throws an error for selected options that aren't present as options (children). Maybe we should consider both the Transfer and the Select in looking at this issue, for consistency.

**Comment by Mohammer5** [comment]
Initially I wanted to introduce a prop that will change the behavior of the Transfer component. But it'd change both the behavior of the component as well as the api surface, which can be confusing and is complicating this already complicated component even further.
I just had a chat with @ismay and we came to the following conclusion:

The requested behavior is already possible with the current  implementation:
* The `selected` options have to be added to the `options` passed to the `Transfer` component
   The selected options wouldn't be displayed by the `Transfer` component as that's handled through internal logic already
* For better UX, the `selected` options should be cached on the apps side and always added to the options
   If the user accidentally deselected an option which hasn't been fetched yet, it'd just completely disappear and couldn't be re-selected
* The app has to ensure that the options passed to the `Transfer` component are unique

**Comment by Mohammer5** [comment]
Here's some pseudo code:

```jsx
const cached = useMemo(() => selected, [])

<Transfer
  options={[...options, ...cached]}
  selected={selected.map(({ value }) => value}
  ...
/>
```

**Comment by ismay** [comment]
Yeah, so what @Mohammer5 is illustrating, is that if you already have the selected options' `labels` and `values`, you might as well just push them to `options`. The component will then just display the selected options. That's the simplest solution, and keeps the component generic as well.

**Comment by Mohammer5** [comment]
Here's a simple implementation of how this could look like in apps:

```jsx
import { PropTypes } from '@dhis2/prop-types'
import { Transfer } from '@dhis2/ui'
import React from 'react'

const DistinctListsTransfer = ({
  options,
  selected,
  ...props,
}) => (
  <Transfer
    {...props}
    options={[...options, ...selected]}
    selected={selected.map(({ value }) => value)}
  />
)

DistinctListsTransfer.propTypes = {
  ...Transfer.propTypes,
  selected: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
    })
  )
}
```

**Comment by martinkrulltott** [comment]
This seems like a reasonable workaround for 2.35. I've implemented it in https://github.com/dhis2/analytics/pull/511 and it seems to work for the period selector 👍 
As the use-case for the period selector is fairly simple (all periods are kept in Analytics, nothing to fetch from the API) I'm not yet sure how this will work in our more advanced use-cases, specifically the Data dimension selector (https://jira.dhis2.org/browse/DHIS2-7943), which loads items dynamically from the API and has a lot more items and filters. However, as the Data dim selector isn't included in 2.35, we can revisit this issue at a later time if needed.

This also adds further to the point in https://github.com/dhis2/ui/issues/63 to include the loading spinner for both sides of the Transfer component, as there's now an even stronger bond between the two sides.

**Comment by Mohammer5** [comment]
I'll close this issue as we've established a working workaround. Loading states for both lists have beed added already
